Sources/WalletConnectAuth/Engine/Common/SessionEngine.swift

```diff
@@ -139,34 +165,23 @@ final class SessionEngine {
             }
         }.store(in: &publishers)
     }
-
-    func settle(topic: String, proposal: SessionProposal, namespaces: [String: SessionNamespace]) throws {
-        try Validator.validate(namespaces) // FIXME: Validation should happen before responding proposal, before settlement
-        let agreementKeys = try! kms.getAgreementSecret(for: topic)!
-        
-        let selfParticipant = Participant(publicKey: agreementKeys.publicKey.hexRepresentation, metadata: metadata)
-        
-        let expectedExpiryTimeStamp = Date().addingTimeInterval(TimeInterval(WCSession.defaultTimeToLive))
-        guard let relay = proposal.relays.first else {return}
-        let settleParams = SessionType.SettleParams(
-            relay: relay,
-            controller: selfParticipant,
-            namespaces: namespaces,
-            expiry: Int64(expectedExpiryTimeStamp.timeIntervalSince1970))//todo - test expiration times
-        let session = WCSession(
-            topic: topic,
-            selfParticipant: selfParticipant,
-            peerParticipant: proposal.proposer,
-            settleParams: settleParams,
-            acknowledged: false)
-        logger.debug("Sending session settle request")
-        Task { try? await networkingInteractor.subscribe(topic: topic) }
-        sessionStore.setSession(session)
-        networkingInteractor.request(.wcSessionSettle(settleParams), onTopic: topic)
-    }
-
+    
     private func onSessionSettle(payload: WCRequestSubscriptionPayload, settleParams: SessionType.SettleParams) {
         logger.debug("Did receive session settle request")
+        guard let proposedNamespaces = settlingProposal?.requiredNamespaces else {
```

Comment:
1. you set `settlingProposal` `onApprovalResponse` so it happens on `responder's` engine as responder calls "approve"
2. you retrieve that `settlingProposal` `onSessionSettle` so it happens on `proposer's` engine as responder calls  "settle"

🤔  am I correct? will it work?

Reply:
splitting engines/methods for controller/non-controller later will really help.